code/ecdsap256sha2/Hacl.Impl.ECDSA.P256SHA256.Verification.Agile.fst

```diff
@@ -88,29 +95,50 @@ let ecdsa_verification_step1 r s =
   let isSCorrect = isMoreThanZeroLessThanOrderMinusOne s in
   isRCorrect && isSCorrect
 
-
-inline_for_extraction noextract
-val ecdsa_verification_step23: hashAsFelem : felem -> mLen: size_t -> m: lbuffer uint8 mLen -> Stack unit
-  (requires fun h -> live h hashAsFelem /\ live h m)
-  (ensures  fun h0 _ h1 -> modifies (loc hashAsFelem) h0 h1 /\
-  (
+inline_for_extraction
+val ecdsa_verification_step23: alg: hash_alg {SHA2_256? alg \/ SHA2_384? alg \/ SHA2_512? alg (* and blake one day *)}
+  ->  mLen: size_t -> m: lbuffer uint8 mLen -> result: felem -> Stack unit
+  (requires fun h -> live h m /\ live h result )
+  (ensures fun h0 _ h1 -> modifies (loc result) h0 h1 /\
+    (
       assert_norm (pow2 32 < pow2 61);
-      let hashM = H.hash Def.SHA2_256 (as_seq h0 m) in 
-      as_nat h1 hashAsFelem == nat_from_bytes_be hashM % prime_p256_order)
+      assert_norm (pow2 32 < pow2 125);
+      let hashM = (hashSpec alg) (as_seq h0 m) in 
+      let cutHashM = Lib.Sequence.sub hashM 0 32 in 
+      as_nat h1 result = nat_from_bytes_be cutHashM % prime_p256_order
+    )
   )
 
-let ecdsa_verification_step23 hashAsFelem mLen m =
-  push_frame();
-        assert_norm (pow2 32 < pow2 61);
-  let h0 = ST.get() in
-  let mHash = create (size 32) (u8 0) in
-  hash_256 m mLen mHash;
-  toUint64ChangeEndian mHash hashAsFelem;
-  let h1 = ST.get() in
-  lemma_core_0 hashAsFelem h1;
-  reduction_prime_2prime_order hashAsFelem hashAsFelem;
-  Spec.ECDSA.changeEndianLemma (uints_from_bytes_be #U64 #_ #4 (as_seq h1 mHash));
-  uints_from_bytes_be_nat_lemma #U64 #_ #4 (as_seq h1 mHash);
+
+let ecdsa_verification_step23 alg mLen m result = 
```

Comment:
this function is exactly the same as `ecdsa_signature_step12` -- please eliminate the code duplication

if the function needs to be duplicated in the C code (which I would not recommend), do:
```
inline_for_extraction
let shared_func () x y z =
  ...

// MyFile.fst
let first_copy = shared_func ()

// MyOtherFile.fst
let second_copy = shared_func ()
```